### ISSUE TITLE
Not getting the same predictions between YOLOv5 predict and pytorch hub 

### ISSUE THREAD

**Issue by anastasia-enot** [ultralytics/yolov5#12121]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello everyone, I have a question about inference from a custom YOLOv5 model.
I have trained it on an image **classification** dataset (binary). I would like to load it using torch.hub.load and to make inference for new images. However, when I do that, I do not obtain the same predictions as using predict. I have already seen the [issue here](https://github.com/ultralytics/yolov5/issues/9030), but even after modifying the code I still do not get the same results. 

Here is the line of code I use for predict:
`python classify.predict --weights runs/train-cls/exp10/weights/best.pt --source PATH --save-txt`

And here is the code for loading the model with Pytorch:

```
model = torch.hub.load('ultralytics/yolov5','custom', path='trained_models/best10.pt', force_reload=True)

IMAGENET_MEAN = 0.485, 0.456, 0.406
IMAGENET_STD = 0.229, 0.224, 0.225

def classify_transforms(size=224): 
     return T.Compose([T.ToTensor(), T.Resize(size), T.CenterCrop(size), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])

images = [os.path.join(path, file) for file in os.listdir(path) if file.endswith('.jpg')]
for img in images:
    image = Image.open(img)
    transformations = classify_transforms()
    convert_tensor = transformations(image)
    convert_tensor = convert_tensor.unsqueeze(0)
    
    results = model(convert_tensor)
    probabilities = torch.nn.functional.softmax(results, dim=1)
    print(probabilities)
```

For example, for the same image with predict I get 0.63 and 0.37 as probabilities of the 2 classes; and with Pytorch I get 0.56921 and 0.43079. 
Could someone pinpoint where I am making a mistake? Any help would be gladly appreciated!

**Comment by glenn-jocher** [comment]
@anastasia-enot hi there! It seems like you're experiencing differences in predictions between using YOLOv5 `predict` and loading the model with `torch.hub.load`. Let's try to identify the issue.

First, make sure you're using the same model weights for both methods. Double-check that the `best.pt` file you're providing to `torch.hub.load` is the same as the one used for `predict`.

Next, there might be differences in data preprocessing between the two approaches. In your `classify_transforms` function, you're applying resizing, center cropping, and normalization. Confirm that these transformations match the ones used during training and inference in the `classify.predict` script.

Additionally, if your previous modifications following issue #9030 didn't resolve the discrepancies, please revert back to the original code and try again, ensuring a clean comparison.

If you're still facing issues after verifying the above points, please provide further details like the exact versions of YOLOv5 and PyTorch you're using, and any additional steps or modifications you made. This will help us assist you more effectively.

Remember, the YOLOv5 community and the Ultralytics team are incredibly helpful and always ready to support you. Feel free to reach out if you have any further questions.

**Comment by anastasia-enot** [comment]
Hello @glenn-jocher , thank you for your answer!
- I did double check that the weights are the same for both methods, so it was not the reason for the problem.
- I feel like it's indeed a problem in data preprocessing but I don't see where I get it wrong. In the YOLOv5 code in utils/augmentations.py this is the following code for transformations: 

```
def classify_transforms(size=224):
    # Transforms to apply if albumentations not installed
    assert isinstance(size, int), f'ERROR: classify_transforms size {size} must be integer, not (list, tuple)'
    # T.Compose([T.ToTensor(), T.Resize(size), T.CenterCrop(size), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])
    return T.Compose([CenterCrop(size), ToTensor(), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])
```

I replaced this in my Pytorch code (I had resize before), it did modify the predictions but it's still not the same as predict.py... 
I will do as you said, start over with a clean repo to see if it changes something.

**Comment by glenn-jocher** [comment]
@anastasia-enot hello! You're welcome, and I'm happy to help.

Based on the code you provided, it seems like the data preprocessing in `classify_transforms` is slightly different from the one used in `predict.py`. In `classify_transforms`, you're applying a center crop instead of a resize operation.

To make the data preprocessing in your PyTorch code match the one used in `predict.py`, you can modify the `classify_transforms` function as follows:

```python
def classify_transforms(size=640):
    return T.Compose([
        T.Resize(size),
        T.CenterCrop(size),
        T.ToTensor(),
        T.Normalize(IMAGENET_MEAN, IMAGENET_STD)
    ])
```

By applying the resize operation before the center crop, you ensure that the size of the input image matches the one expected by the model during training. This change should help align the predictions between `predict.py` and your PyTorch code.

After making this modification, try running your code again and see if the predictions become consistent. Starting with a clean repository is also a good idea to ensure there are no lingering modifications affecting the results.

If you still encounter any issues or have further questions, please don't hesitate to ask. The YOLOv5 community and the Ultralytics team are here to assist you.
